### PR TITLE
[REVIEW] Feature/optimize accessor copy

### LINKED ISSUE #7646: Unsafe usage of OrderedDict inheritance
### Summary

The `OrderedColumnDict` class used as a base for `ColumnAccessor is making some unsafe choices that are not currently leading to bugs, but could lead to developer error in the future. We can either restrict usage explicitly to stop developers from using unsafe APIs, in which case we can make some performance improvements, or we can make a more flexible but slower choice.

### Background

Inheriting from built-in classes in Python can be dangerous because internal optimizations may have unexpected consequences. In this case, the most relevant issue is that overriding `__setitem__` does not necessarily affect other methods in the way that you might expect; in particular, `update` is not guaranteed to call `__setitem__`, but may instead inline the code for performance. Typically, dict-like classes are instead advised to extend `collections.abc.MutableMapping` instead, because in that case implementing the abstract methods _is_ guaranteed to give the expected behavior for the other mixin methods.

The `OrderedColumnDict` currently inherits from `OrderedDict`. In Python 2.7, `OrderedDict` was a standalone class, but in Python 3 [it is a direct subclass of `dict`](https://github.com/python/cpython/blob/master/Lib/collections/__init__.py#L78). However, this is where things get more complicated. `OrderedDict` was implemented in Python until Python 3.5, when a [C implementation was added](https://bugs.python.org/issue16991). Interestingly, while the original Python implementation does not override `update`, the C implementation [hooks back into the `MutableMapping` C implementation](https://github.com/python/cpython/blob/226a012d1cd61f42ecd3056c554922f359a1a35d/Objects/odictobject.c#L1213). This means that up until Python 3.5, overriding `__setitem__` would not change the behavior of `update`, but afterwards it would. 

### Proposal

The upshot is that the `OrderedColumnDict` is currently relying on an implementation detail to make methods like `update` and `setdefault` use the `__setitem__` implementations of `EqualLengthValuesMappingMixin` and `ColumnValuesMappingMixin`. Since CPython makes no guarantees that this behavior will not change in the future, we have two options:

**Option 1**

Instead of inheriting directly from `OrderedDict`, we can implement a `MutableMapping` that owns an `OrderedDict` internally and override `__setitem__` there to perform the necessary conversions. This change will slow the code somewhat, but it will make the object safe to future modifications in the Python code base.

**Option 2**

As far as I can tell, this class is only ever used internally, and only by the `ColumnAccessor`. The closest thing to an exception is that it's constructed [in a temporary table for the purpose of hashing](https://github.com/rapidsai/cudf/blob/branch-0.19/python/cudf/cudf/core/dataframe.py#L4857), and even then it's internally being passed to the `ColumnAccessor` constructor. That being the case, we could simply override all other mutating methods of `OrderedColumnDict` (of the top of my head that's just `update` and `setdefault`), either to raise exceptions so that future developers can't accidentally use them, or to implement them by calling `__setitem__`. If it's never a user-facing object, either change seems like a reasonable choice. The important thing is to either 1) ensure that all methods that we use are correctly overridden, or 2) only ever use `__setitem__` so that we don't rely on any other behavior.

**Performance Implications**

I originally started looking at this code while playing with some benchmarking code from @shwina. For data frames with 1M rows, the shallow copying of columns that happens in order to match dtypes (see `Merge._match_key_dtypes` in `cudf/core/join.py`) takes around ~5% of the execution time of the entire merge by itself, and that jumps into the 10-15% range for data frames with 1000 rows. 

In either option, benchmarks suggest that we would benefit from inlining the `__setitem__` logic from the mixins. `__setitem__` calls are not free, especially once you're overriding the C implementations of built-in (even a trivial implementation that just calls `super().__setitem__(...)` is close to an order of magnitude slower), so we would benefit measurably from avoiding the extra Python function call overhead and MRO lookup by just combining the mixin implementations of `__setitem__` into a single `__setitem__` override in `OrderedColumnDict`.

Option 2 would also enable some further optimization. Since `dict` is insertion-ordered as of Python 3.7, we could safely have `OrderedColumnDict` inherit from `dict` instead of `OrderedDict`, which would not only make `__setitem__` calls about 10% faster according to my benchmarks, but make copies about twice as fast, which in turn would make shallow copying columns much faster, an operation that seems to occur fairly often due to the need to match dtypes in join-like operations.

### PR DESCRIPTION
**Pull request by vyasr**
This PR makes a number of changes to resolve #7646:
1. The underlying data store in `ColumnAccessor` has been converted to a `dict`. The conversion from an `OrderedDict` subclass to a raw `dict` is safe because, as of Python 3.7, dictionary ordering is guaranteed by the CPython standard.
2. The validation previously defined by a pair of mixin classes that were parents of the `OrderedColumnDict` class has instead been moved to a new method that is called in `ColumnAccessor.set_by_label`. This ensures that the validation only happens once, and dramatically speeds up most dict operations since they can now rely on the built-in C implementation of `__setitem__` rather than Python overrides. Furthermore, this obviates the safety concerns raised in #7646 since we are no longer overriding these methods.
3. The constructor of `ColumnAccessor` has been carefully optimized to handle validation with minimal overhead. Previously, this was handled by `_data.__setitem__`.
4. The column length in a `ColumnAccessor` is stored the first time a column is added so that it doesn't have to be recomputed.
4. A new parameter `validate` has been added to `_set_by_label`. This parameter is currently unused, but can be used in future refactorings to further speed up the code by bypassing validation steps when we are internally passing columns that we know to be safe.

The performance implications of these changes for shallow copying `DataFrame` objects are shown below. Specifically, I'm benchmarking `df.copy(deep=False)` for different numbers of columns and column sizes. As expected, these changes are largely invariant to the size of the columns and scale with the number of columns. The scaling isn't perfectly linear because of the constant overhead of copying the `ColumnAccessor` itself, but the speedup plateaus around 5x for >200 columns (which probably isn't too realistic anyway). Exactly how these speedups will propagate to operations like joins remains to be seen, but this should help since some crude benchmarking I did showed that copying can take up to 15-20% of the total time of a join depending on the data size. @shwina has better benchmarks there and I'll work with him on future changes. [Here's the benchmarking notebook I used](https://github.com/rapidsai/cudf/files/6174381/CopyDfBmark.txt). Note that since Github doesn't support uploading notebooks I've changed the extension to txt, but it will work find to just rename it to a .ipynb file.

image

### CODE REVIEW

**Review by shwina — APPROVED** [review]
LGTM! Nice work @vyasr!

**Review by isVoid — APPROVED** [review]
Awesome!

**Review by kkraus14 — APPROVED** [review]

File: python/cudf/cudf/core/column_accessor.py [hunk 1]
```diff
@@ -270,16 +296,29 @@ def select_by_index(self, index: Any) -> ColumnAccessor:
...
+        validate : bool
+            If True, the provided value will be coerced to a column and
+            validated before setting (Default value = True).
         """
         key = self._pad_key(key)
+        if validate:
+            value = column.as_column(value)
+            if len(self._data) > 0:
+                if len(value) != self._column_length:
+                    raise ValueError("All columns must be of equal length")
```
Comment (shwina) [inline python/cudf/cudf/core/column_accessor.py:318]: Looks like we changed the error message here -- we should fix the corresponding test in https://github.com/rapidsai/cudf/blob/d500142e91c041a69b531d06f346e1630a8af046/python/cudf/cudf/tests/test_dataframe.py#L5225

Comment (kkraus14) [reply python/cudf/cudf/core/column_accessor.py:318]: I think we changed the other error messages back to what they were

File: python/cudf/cudf/core/column_accessor.py [hunk 2]
```diff
@@ -270,16 +296,29 @@ def select_by_index(self, index: Any) -> ColumnAccessor:
...
+        key
+            name of the column
         value : column-like
+            The value to insert into the column.
+        validate : bool
+            If True, the provided value will be coerced to a column and
+            validated before setting (Default value = True).
         """
         key = self._pad_key(key)
+        if validate:
```
Comment (shwina) [inline python/cudf/cudf/core/column_accessor.py:314]: We need to do similar validation in `insert` as well.

File: python/cudf/cudf/core/column_accessor.py (hunk 1 above)
Comment (vyasr) [reply python/cudf/cudf/core/column_accessor.py:318]: I made this change yesterday, looks like I missed one. I think I got them all this time, tests are passing locally.

**Comment by kkraus14** [comment]
@gpucibot merge

**Pull request merged**